### PR TITLE
add atomic unsafe response type. enable async handler

### LINKED ISSUE #308: Use of actor reference on Async Actor handle
I'm trying implement some asynchronous actor response that uses the Actor's self reference on _actix v0.9_ but I can't get it to work on the account of some lifetime bounds error. I've separated a small example that contains the error:

```Rust
use actix::prelude::*;

#[derive(Message)]
#[rtype(result = "Result<usize, ()>")]
struct Msg;

struct MyActor;

impl Actor for MyActor {
    type Context = Context<Self>;
}

impl Handler<Msg> for MyActor {
    type Result = ResponseActFuture<Self, Result<usize, ()>>;

    fn handle(&mut self, _msg: Msg, _: &mut Context<Self>) -> Self::Result {
        Box::new(self.reply().into_actor(self))
    }
}

impl MyActor {
    async fn reply(&self) -> Result<usize, ()> {
        Ok(42)
    }
}
```
But this gives me the following error:
```
error[E0495]: cannot infer an appropriate lifetime for autoref due to conflicting requirements
  --> src/main.rs:17:23
   |
17 |         Box::new(self.reply().into_actor(self))
   |                       ^^^^^
   |
note: first, the lifetime cannot outlive the anonymous lifetime #1 defined on the method body at 16:5...
  --> src/main.rs:16:5
   |
16 | /     fn handle(&mut self, _msg: Msg, _: &mut Context<Self>) -> Self::Result {
17 | |         Box::new(self.reply().into_actor(self))
18 | |     }
   | |_____^
note: ...so that reference does not outlive borrowed content
  --> src/main.rs:17:18
   |
17 |         Box::new(self.reply().into_actor(self))
   |                  ^^^^
   = note: but, the lifetime must be valid for the static lifetime...
note: ...so that the expression is assignable
  --> src/main.rs:17:9
   |
17 |         Box::new(self.reply().into_actor(self))
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   = note: expected  `std::boxed::Box<(dyn actix::fut::ActorFuture<Actor = MyActor, Output = std::result::Result<usize, ()>> + 'static)>`
              found  `std::boxed::Box<dyn actix::fut::ActorFuture<Actor = MyActor, Output = std::result::Result<usize, ()>>>`
```

I've tried to implement this using the `ActorFuture` and the `them` combinator but I get similar errors, this is what I tried:
```Rust
    fn handle(&mut self, _msg: Msg, _: &mut Context<Self>) -> Self::Result {
        Box::new(
            async {}
                .into_actor(self)
                .then(|_, act, _| actix::fut::wrap_future(act.reply())),
        )
    }
```

The problem seams to be that I can only have static references for the future I return, but since the future will be handled by the Actor itself shouldn't it be able to use it's own  reference?

Are there examples for asynchronously handling messages? I have only been able to find the doc example on `ActorFuture`, witch is even a little bit outdated; but it's a bit of a different context since that doesn't use a reference to the actor on the futures it chains.

### PR DESCRIPTION
**Pull request by fakeshadow**
## PR Type

Feature

## PR Checklist
Check your PR fulfills the following:

- [x] Tests for the changes have been added / updated.
- [x] Documentation comments have been added / updated.
- [x] A changelog entry has been made for the appropriate packages.
- [x] Format code with the latest stable rustfmt

## Overview

AtomicResponse actually could enable borrowing Actor and it's Context in async/await.
This PR just use it with some unsafe code to enable async/await.

It's also possible to immutably borrowing Actor and it's Context in async/await without atomic,
by grouping shared reference responses and block actor on them.
It's basically a batched version of AtomicResponse. 

Closes #308 #378 #438

### CODE REVIEW

**Comment by fakeshadow** [comment]
I'm not yet decided how to guard the input. 
Introducing async_trait for a future proof of Handler trait is what I prefer but it could mean harm to SyncActor usage.
Use an extra closure would not break but with more boilerplate

**Comment by robjtede** [comment]
So what's the path forward for this PR. Is it ready for review and if so which parts should be focussed on.

A note after skimming over a couple files: a few fns are documented with "do not expose to public api". Are these fns better as unsafe? Also consider documenting the invariants that must be upheld regardless.

**Comment by fakeshadow** [comment]
Hard to say it's ready as it's pretty hard to test the unsafes added. It would need some really edge cases to see if it holds up.

`Context::wait_concurrent` is just a method to push 'static lifetime future to actor's context. It's like `actix_rt::spawn` on an actor and they are safe. It' just not suggested to be used in public API. 

The unsafe code only affect people who actually call `ResponseAsync` in handler. For people avoid using it there is no actual change to the code base other than a few more `Vec::len` check against an empty vector.

**Comment by fakeshadow** [comment]
Making this a draft for now until proper tests are added.

**Comment by fakeshadow** [comment]
I have reduced most changes related to actor context. Changing this PR to an `AtomicResponse` similar feature that uses unsafe to permit borrowing actor and context in async block. 
It's supposed to work exactly like `AtomicResponse` but async/await friendly.

**Comment by robjtede** [comment]
definitely excited about this but i think we should do a release without it first

**Comment by fakeshadow** [comment]
Agreed. I'm not confident in this either.

There are still small issues to this approach so it's more like an experiment for the Concurrent variant AsyncResponse.

**Comment by fakeshadow** [comment]
Current issue I find on this approach:

`Actor::stopping` have access to actor and context with mut reference. So an actor with living waiting async response would find the state could be modified between yield points where the actor goes through the stopping and back to running state with modifed state. This is not a double mut borrow by any means but would still be annoying as the actor state loses consistency between await point when above situation happen.

The only way to fix this is to change the current behavior of handling stopping or use a spearate queue for async responses where they are thrown away when actor goes into stopping state.

**Comment by robjtede** [comment]
lmk if you want a review of this now

**Comment by fakeshadow** [comment]
@robjtede I think this is ready for review.

**Comment by fakeshadow** [comment]
I don't have time to keep this PR up to date anymore. It's welcomed to step in and take charge of it.

File: actix/src/handler.rs [hunk 1]
```diff
@@ -151,6 +151,93 @@ where
...
+where
+    A: Actor,
+    A::Context: AsyncContext<A>,
+    M: Message<Result = T>,
+    T: 'static,
+{
+    fn handle(self, ctx: &mut A::Context, tx: Option<OneshotSender<M::Result>>) {
+        match self {
+            AsyncResponse::Atomic(fut) => {
+                ctx.wait(fut.map(move |res, _, _| tx.send(res)));
```
Comment (thalesfragoso) [inline actix/src/handler.rs:232]: Sorry for reviving this....

Isn't this unsound already ? You have two `&mut A::Context` alive at the same time, one given to the method and the other inside the future.
